statsmodels/regression/linear_model.py

```diff
@@ -330,6 +332,40 @@ def fit(self, method="pinv", cov_type='nonrobust', cov_kwds=None,
             # used in ANOVA
             self.effects = effects = np.dot(Q.T, self.wendog)
             beta = np.linalg.solve(R, effects)
+        elif method == "qr-pivot":
+            Q, R, P = linalg.qr(self.wexog, mode='economic', pivoting=True)
+            self.exog_Q, self.exog_R, self.exog_P = Q, R, P
+
+            # Cache singular values from R.
+            self.wexog_singular_values = np.linalg.svd(R, 0, 0)
+            if 'tol' in kwargs:
+                tol = kwargs['tol']
+            else:
+                tol = 1e-7
+
+            singular_preds = abs(np.diagonal(R)) < tol
+            n_singular = sum(singular_preds)
+            n_preds = len(singular_preds)
+            self.rank = n_preds - n_singular
+            sortP = np.argsort(P)
+
+            if n_singular > 0:
+                rng = range(self.rank)
+                Qq = Q[:, rng]
+                Rr = R[rng, :][:, rng]
+                effects = np.dot(Qq.T, self.wendog)
+                beta = linalg.solve_triangular(Rr, effects)
+                beta = np.concatenate((beta, np.repeat(0.0, n_singular)))[sortP]
+                self.effects = np.concatenate((effects, np.repeat(np.NaN, n_singular)))
+                normalized_cov_params = np.linalg.inv(np.dot(Rr.T, Rr))
+
+                normalized_cov_params = np.vstack((normalized_cov_params, np.zeros((n_singular, self.rank))))
+                normalized_cov_params = np.hstack((normalized_cov_params, np.zeros((n_preds, n_singular))))
+                self.normalized_cov_params = normalized_cov_params[sortP, :][:, sortP]
+            else:
+                self.effects = np.dot(Q.T, self.wendog)
+                beta = linalg.solve_triangular(R, self.effects)[sortP]
```

Comment:
Actually there is only one sort and the algorithm only apply the permutation multiple times. 

The reason it’s using the “triagonal solve” for the linear equations instead of rearranging the R matrix and solve the linear equations using the invers matrix of R. Triagonal solve is more efficient (O(n^2)) and more accurate compared to taking the inverse of a matrix O(n^3) (or close to it). 

Applying the sort permutation on the way it is may looks more complex but overall it solves the linear equations with complexity of O(n^2) and offers better accuracy.